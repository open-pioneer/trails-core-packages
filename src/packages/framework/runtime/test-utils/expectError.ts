export function expectError(impl: () => unknown): Error {
    try {
        impl();
        throw new Error("expected error!");
    } catch (e) {
        if (e instanceof Error) {
            return e;
        }
        throw new Error("unexpected error value, not an instance of Error");
    }
}

export function expectAsyncError(impl: () => Promise<unknown>): Promise<Error> {
    const promise = impl();
    return promise.then(
        () => {
            throw new Error("expected error!");
        },
        (e) => {
            if (e instanceof Error) {
                return e;
            }
            throw new Error("unexpected error value, not an instance of Error");
        }
    );
}
