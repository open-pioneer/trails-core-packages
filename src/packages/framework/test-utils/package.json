{
    "name": "@open-pioneer/test-utils",
    "private": true,
    "dependencies": {
        "@testing-library/dom": "^9.0.0",
        "@formatjs/intl": "^2.6.7"
    },
    "peerDependencies": {
        "@open-pioneer/runtime": "workspace:^"
    },
    "devDependencies": {
        "@open-pioneer/runtime": "workspace:^",
        "@testing-library/react": "^14.0.0",
        "react": "^18.2.0",
        "react-dom": "^18.2.0"
    }
}
