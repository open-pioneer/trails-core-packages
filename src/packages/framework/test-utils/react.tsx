import { createIntl, createIntlCache, IntlShape } from "@formatjs/intl";
import { Service } from "@open-pioneer/runtime";
import {
    PackageContext as InternalPackageContext,
    PackageContextMethods
} from "@open-pioneer/runtime/react-integration";
import { FC, ReactNode, useMemo } from "react";

export type AnyService = Service<Record<string, unknown>>;

export interface PackageContextProviderProps {
    /** Interface implementations, keyed by interface name. */
    services?: {
        [interfaceName: string]: AnyService;
    };

    /** Interface implementations, keyed by interface name and then by qualifier. */
    qualifiedServices?: {
        [interfaceName: string]: {
            [qualifier: string]: AnyService;
        };
    };

    /** Package properties (keyed by package name). */
    properties?: {
        [packageName: string]: Record<string, unknown>;
    };

    /**
     * The locale for i18n messages and formatting.
     *
     * @default "en"
     */
    locale?: string;

    /**
     * I18n messages for packages
     */
    messages?: {
        /**
         * I18n messages as (messageId, message) entries.
         *
         * @default {}
         */
        [packageName: string]: Record<string, string>;
    };

    /** Children to render */
    children?: ReactNode;
}

/**
 * Injects services and properties into the component tree.
 * React components using pioneer hooks like `useService` and `useProperties`
 * will receive the mocked properties here instead.
 */
export const PackageContextProvider: FC<PackageContextProviderProps> = (props) => {
    const { children, ...rest } = props;
    const contextMethods = useMemo(() => createPackageContextMethods(rest), [rest]);
    return (
        <InternalPackageContext.Provider value={contextMethods}>
            {children}
        </InternalPackageContext.Provider>
    );
};

function createPackageContextMethods(
    options: Omit<PackageContextProviderProps, "children">
): PackageContextMethods {
    const services = options?.services ?? {};
    const qualifiedServices = options?.qualifiedServices ?? {};
    const properties = options?.properties ?? {};
    const locale = options?.locale ?? "en";
    const messages = options?.messages ?? {};
    const cachedIntl: Record<string, IntlShape> = {};
    return {
        getService(packageName, interfaceName, options) {
            if (!options.qualifier) {
                const service = services[interfaceName];
                if (service) {
                    return service;
                }
                const qualified = qualifiedServices[interfaceName];
                if (qualified) {
                    for (const value of Object.values(qualified)) {
                        return value; // Return any implementation
                    }
                }
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}'. Update the configuration of PackageContextProvider.`
                );
            }

            const qualifiedService = qualifiedServices[interfaceName]?.[options.qualifier];
            if (!qualifiedService) {
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}' (qualifier '${options.qualifier}'). Update the configuration of PackageContextProvider.`
                );
            }
            return qualifiedService;
        },
        getServices(packageName, interfaceName) {
            const unqualified = services[interfaceName];
            const results = Object.values(qualifiedServices[interfaceName] ?? {});
            if (unqualified) {
                results.push(unqualified);
            }
            return results;
        },
        getProperties(packageName) {
            const packageProperties = properties[packageName];
            if (!packageProperties) {
                throw new Error(
                    `No properties for package '${packageName}' bound for testing. Update the configuration of PackageContextProvider.`
                );
            }
            return packageProperties;
        },
        getIntl(packageName) {
            const initIntl = () => {
                const packageMessages = messages[packageName];
                const cache = createIntlCache();
                return createIntl(
                    {
                        locale,
                        messages: packageMessages
                    },
                    cache
                );
            };
            return (cachedIntl[packageName] ??= initIntl());
        }
    };
}
