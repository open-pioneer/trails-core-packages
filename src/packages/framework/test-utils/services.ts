// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { ReferenceMeta, Service, ServiceConstructor } from "@open-pioneer/runtime";
import { createIntl, createIntlCache } from "@formatjs/intl";

/**
 * Options for the {@link createService} function.
 */
export interface CreateServiceOptions<References> {
    /**
     * References provided to the service constructor.
     *
     * @default {}
     */
    references?: PartialServiceReferences<References>;

    /**
     * Properties provided to the service constructor.
     *
     * @default {}
     */
    properties?: Record<string, unknown>;

    /**
     * The locale for i18n messages and formatting.
     *
     * @default "en"
     */
    locale?: string;

    /**
     * The locale for embedded default messages.
     *
     * See also https://formatjs.io/docs/intl#message-descriptor
     *
     * @default "en"
     */
    defaultMessageLocale?: string;

    /**
     * I18n messages as (messageId, message) entries.
     *
     * @default {}
     */
    messages?: Record<string, string>;
}

type PartialServiceReferences<References> = {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [referenceName in keyof References]?: References[referenceName] extends any[]
        ? Partial<References[referenceName][number]>[]
        : Partial<References[referenceName]>;
};

/**
 * Creates a new service instance of `clazz`.
 *
 * The options passed into the constructor of `clazz` are taking from the `options` parameter.
 *
 * This function allows the user to define mock- or test objects as service references.
 *
 * Most type-checking features are disabled for convenience, but by default this function's types
 * enforce that a value provided for a reference must match the partial interface of the actually required type.
 *
 * For example, if a service depends on `{ ctx: ApplicationContext }`, this function ensures
 * that the test value for `ctx` implements `Partial<ApplicationContext>`.
 * This can of course be disabled by casting the value explicitly.
 *
 * @param clazz a service class
 * @param options options passed to the service constructor
 * @returns a new instance of the given service class
 */
export async function createService<References extends {}, Interface extends {}>(
    clazz: ServiceConstructor<References, Interface>,
    options?: CreateServiceOptions<References>
): Promise<Service<Interface>> {
    const references = options?.references ?? {};
    const referencesMeta = Object.fromEntries(
        Object.entries(references).map(([referenceName, reference]) => [
            referenceName,
            referenceMeta(referenceName, reference)
        ])
    );

    const locale = options?.locale ?? "en";
    const defaultMessageLocale = options?.defaultMessageLocale ?? "en";
    const messages = options?.messages ?? {};
    const cache = createIntlCache();
    const intl = createIntl(
        {
            locale,
            defaultLocale: defaultMessageLocale,
            messages
        },
        cache
    );

    return new clazz({
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        references: references as any,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        referencesMeta: referencesMeta as any,
        properties: options?.properties ?? {},
        intl
    });
}

function referenceMeta(referenceName: string, reference: unknown): ReferenceMeta | ReferenceMeta[] {
    if (Array.isArray(reference)) {
        return reference.map<ReferenceMeta>((_, index) => ({
            serviceId: `test-utils::${referenceName}-${index}`
        }));
    }
    return {
        serviceId: `test-utils::${referenceName}`
    };
}
