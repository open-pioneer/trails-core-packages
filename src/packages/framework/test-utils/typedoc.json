{
    "extends": ["../../../../typedoc.base.json"],
    "entryPoints": ["./web-components.ts", "./react.tsx", "./services.ts"]
}
