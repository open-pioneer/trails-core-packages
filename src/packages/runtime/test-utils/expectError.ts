// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
export function expectError(impl: () => unknown): Error {
    try {
        impl();
        throw new Error("expected error!");
    } catch (e) {
        if (e instanceof Error) {
            return e;
        }
        throw new Error("unexpected error value, not an instance of Error");
    }
}

export function expectAsyncError(impl: () => Promise<unknown>): Promise<Error> {
    const promise = impl();
    return promise.then(
        () => {
            throw new Error("expected error!");
        },
        (e) => {
            if (e instanceof Error) {
                return e;
            }
            throw new Error("unexpected error value, not an instance of Error");
        }
    );
}
