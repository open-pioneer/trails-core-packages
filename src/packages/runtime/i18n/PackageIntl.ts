// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import {
    createIntl,
    createIntlCache,
    IntlConfig,
    IntlErrorCode,
    IntlFormatters,
    IntlShape,
    MessageDescriptor,
    OnErrorFn
} from "@formatjs/intl";
import {
    type FormatXMLElementFn,
    type Options as IntlMessageFormatOptions,
    type PrimitiveType
} from "intl-messageformat"; // not a dependency, this is a dependency of formatjs
import { createElement, Fragment, ReactNode, Children } from "react";

type BaseIntl = Pick<IntlShape, "locale" | "timeZone"> & IntlFormatters<string>;

/**
 * Gives access to the package's i18n messages for the current locale.
 *
 * See also https://formatjs.io/docs/intl
 */
export type PackageIntl = BaseIntl & PackageIntlExtensions;

/**
 * Trails specific extensions to formatjs' intl API.
 */
export interface PackageIntlExtensions {
    /**
     * Similar to `formatMessage()`, but supports _rich text formatting_.
     *
     * React nodes are supported as input values, and custom tags can be implemented as react components.
     * Note that the output is always a single react node.
     *
     * *Example with react node as value:*
     *
     * ```tsx
     * function Example() {
     *   // Given i18n message "Hello, {name}!"
     *   // replaces 'name' with the given react node:
     *   const message = intl.formatRichMessage({ id: "foo"}, {
     *     name: <FancyUserName />
     *   });
     *   return <Box>{message}</Box>;
     * }
     * ```
     *
     * *Example with basic formatting:*
     *
     * ```tsx
     * function Example() {
     *   // Given i18n message "Hello, <strong>{name}</strong>!"
     *   // renders with actual <strong> html node:
     *   return <Box>{intl.formatRichMessage({ id: "foo"}, { name: "User" })}</Box>;
     * }
     * ```
     *
     * Note that only a few basic formatting tags are predefined ("b", "strong", "i", "em", "code", "br").
     * If you need more advanced tags, you can define your own, see below.
     *
     * *Example with custom tag:*
     *
     * ```tsx
     * function Example() {
     *   // Given i18n message "Open <foo>the door</foo>!",
     *   // renders 'foo' using the formatter function below:
     *   const message = intl.formatRichMessage({ id: "foo"}, {
     *     foo: (parts) => <FancyTag>{parts}</FancyTag>
     *   });
     *   return <Box>{message}</Box>;
     * }
     * ```
     */
    formatRichMessage: (
        descriptor: MessageDescriptor,
        values?: Record<string, RichTextValue>,
        opts?: IntlMessageFormatOptions
    ) => ReactNode;
}

/**
 * Value types supported when rendering rich text messages.
 * Primitive values (or single react nodes) are used for normal values.
 * Functions are used to define how tags should be rendered.
 *
 * See {@link PackageIntlExtensions.formatRichMessage}
 */
export type RichTextValue = PrimitiveType | ReactNode | FormatXMLElementFn<ReactNode>;

/**
 * Creates a package intl instance with the given locale and messages.
 *
 * @internal
 */
export function createPackageIntl(
    locale: string,
    messages: Record<string, string>,
    options?: {
        /** Used during testing to suppress console messages. */
        suppressNotFoundError?: boolean;

        /** Also used during testing */
        defaultLocale?: string;
    }
): PackageIntl {
    const cache = createIntlCache();
    const config: IntlConfig = {
        locale,
        messages
    };
    if (options?.suppressNotFoundError) {
        config.onError = ignoreMissingTranslationError;
    }
    if (options?.defaultLocale) {
        config.defaultLocale = options.defaultLocale;
    }

    const intl = createIntl(config, cache) as unknown as PackageIntl;
    intl.formatRichMessage = formatRichMessage.bind(undefined, intl);
    return intl;
}

/** Creates an empty i18n instance, e.g. for tests. */
export function createEmptyPackageIntl(locale = "en"): PackageIntl {
    return createPackageIntl(locale, {});
}

const RICH_TEXT_TAGS = Object.freeze({
    "b": renderTag("b"),
    "strong": renderTag("strong"),
    "i": renderTag("i"),
    "em": renderTag("em"),
    "code": renderTag("code"),
    "br": renderTag("br")
});

function formatRichMessage(
    intl: PackageIntl,
    descriptor: MessageDescriptor,
    values: Record<string, RichTextValue> | undefined,
    opts: IntlMessageFormatOptions | undefined
): ReactNode {
    values = {
        ...RICH_TEXT_TAGS,
        ...values
    };

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const formatResult: any = intl.formatMessage(descriptor, preprocessValues(values) as any, opts);
    return createElement(Fragment, undefined, ...formatResult);
}

function preprocessValues(values: Record<string, RichTextValue> | undefined) {
    if (!values) {
        return undefined;
    }

    const fixedValues: Record<string, RichTextValue> = {};
    for (const [key, valueOrFn] of Object.entries(values)) {
        let fixedValue = valueOrFn;
        if (typeof valueOrFn === "function") {
            fixedValue = (parts) => toKeyedChildren(valueOrFn(parts));
        }
        fixedValues[key] = fixedValue;
    }
    return fixedValues;
}

function renderTag(tag: string): FormatXMLElementFn<ReactNode> {
    return function renderRichTag(parts) {
        return createElement(tag, undefined, ...parts);
    };
}

// Seems to be necessary to avoid "Each child in a list should have a unique "key" prop." error
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function toKeyedChildren(children: any) {
    return Children.toArray(children);
}

/** Hides missing translation errors during tests */
const ignoreMissingTranslationError: OnErrorFn = (err) => {
    if (err.code === IntlErrorCode.MISSING_TRANSLATION) {
        return;
    }

    console.error(err);
};
