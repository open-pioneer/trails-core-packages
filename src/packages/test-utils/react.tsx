// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import type { PackageIntl, Service } from "@open-pioneer/runtime";
import {
    PackageContext as InternalPackageContext,
    PackageContextMethods
} from "@open-pioneer/runtime-react-support";
import { CustomChakraProvider } from "@open-pioneer/runtime/test-support";
import { FC, ReactNode, useMemo } from "react";
import { createIntl } from "./vanilla";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type AnyService = Service<any>;

export interface PackageContextProviderProps {
    /** Interface implementations, keyed by interface name. */
    services?: {
        [interfaceName: string]: AnyService;
    };

    /** Interface implementations, keyed by interface name and then by qualifier. */
    qualifiedServices?: {
        [interfaceName: string]: {
            [qualifier: string]: AnyService;
        };
    };

    /** Package properties (keyed by package name). */
    properties?: {
        [packageName: string]: Record<string, unknown>;
    };

    /**
     * The locale for i18n messages and formatting.
     *
     * @default "en"
     */
    locale?: string;

    /**
     * I18n messages for packages
     */
    messages?: {
        /**
         * I18n messages as (messageId, message) entries.
         *
         * @default {}
         */
        [packageName: string]: Record<string, string>;
    };

    /** Children to render */
    children?: ReactNode;
}

/**
 * Injects services and properties into the component tree.
 * React components using Open Pioneer Trails hooks like `useService` and `useProperties`
 * will receive the mocked properties here instead.
 * Currently also wrapped with the CustomChakraProvider to support chakra-ui elements
 */
export const PackageContextProvider: FC<PackageContextProviderProps> = (props) => {
    const { children, ...rest } = props;
    const [locale, contextMethods] = useMemo(() => createPackageContextMethods(rest), [rest]);

    return (
        <CustomChakraProvider rootNode={document} appRoot={document.body} locale={locale}>
            <InternalPackageContext.Provider value={contextMethods}>
                {children}
            </InternalPackageContext.Provider>
        </CustomChakraProvider>
    );
};

function createPackageContextMethods(
    options: Omit<PackageContextProviderProps, "children">
): [locale: string, PackageContextMethods] {
    const services = options?.services ?? {};
    const qualifiedServices = options?.qualifiedServices ?? {};
    const properties = options?.properties ?? {};
    const locale = options?.locale ?? "en";
    const messages = options?.messages ?? {};
    const cachedIntl: Record<string, PackageIntl> = {};
    const methods: PackageContextMethods = {
        getService(packageName, interfaceName, options) {
            if (!options.qualifier) {
                const service = services[interfaceName];
                if (service) {
                    return service;
                }
                const qualified = qualifiedServices[interfaceName];
                if (qualified) {
                    for (const value of Object.values(qualified)) {
                        return value; // Return any implementation
                    }
                }
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}'. Update the configuration of PackageContextProvider.`
                );
            }

            const qualifiedService = qualifiedServices[interfaceName]?.[options.qualifier];
            if (!qualifiedService) {
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}' (qualifier '${options.qualifier}'). Update the configuration of PackageContextProvider.`
                );
            }
            return qualifiedService;
        },
        getServices(packageName, interfaceName) {
            const unqualified = services[interfaceName];
            const results = Object.values(qualifiedServices[interfaceName] ?? {});
            if (unqualified) {
                results.push(unqualified);
            }
            return results;
        },
        getProperties(packageName) {
            const packageProperties = properties[packageName];
            if (!packageProperties) {
                throw new Error(
                    `No properties for package '${packageName}' bound for testing. Update the configuration of PackageContextProvider.`
                );
            }
            return packageProperties;
        },
        getIntl(packageName) {
            const initIntl = () => {
                const packageMessages = messages[packageName];
                return createIntl({
                    locale,
                    messages: packageMessages
                });
            };
            return (cachedIntl[packageName] ??= initIntl());
        }
    };
    return [locale, methods];
}
