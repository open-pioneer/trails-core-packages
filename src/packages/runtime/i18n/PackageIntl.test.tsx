// SPDX-FileCopyrightText: 2023-2025 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { Box, Tag } from "@chakra-ui/react";
import { PackageContextProvider } from "@open-pioneer/test-utils/react";
import { render, screen } from "@testing-library/react";
import { ReactNode } from "react";
import { describe, expect, it } from "vitest";
import { createPackageIntl } from "./PackageIntl";

describe("formatRichMessage", () => {
    it("formats messages containing react nodes", async () => {
        const intl = createPackageIntl("en-US", {
            test: "Hello, {name}!"
        });
        const message = intl.formatRichMessage(
            {
                id: "test"
            },
            {
                name: (
                    <Tag.Root>
                        <Tag.Label>User</Tag.Label>
                    </Tag.Root>
                )
            }
        );
        renderBox(message);

        const element = await screen.findByTestId("test");
        expect(element).toMatchInlineSnapshot(`
          <div
            data-testid="test"
          >
            Hello, 
            <div
              class="chakra-tag__root css-fs7q9r"
            >
              <span
                class="chakra-tag__label css-1v61si9"
              >
                User
              </span>
            </div>
            !
          </div>
        `);
    });

    it("formats messages containing basic html tags", async () => {
        const intl = createPackageIntl("en-US", {
            test: "Hello, <strong>{name}</strong>!"
        });
        const message = intl.formatRichMessage(
            {
                id: "test"
            },
            {
                name: "User"
            }
        );
        renderBox(message);

        const element = await screen.findByTestId("test");
        expect(element).toMatchInlineSnapshot(`
          <div
            data-testid="test"
          >
            Hello, 
            <strong>
              User
            </strong>
            !
          </div>
        `);
    });

    it("supports defining custom tags", async () => {
        const intl = createPackageIntl("en-US", {
            test: "Hello, <custom1>Content of custom tag <custom2>and nested tag with {param}</custom2></custom1>!"
        });
        const message = intl.formatRichMessage(
            {
                id: "test"
            },
            {
                custom1: (parts) => {
                    return <Box className="from-custom-1">{parts}</Box>;
                },
                custom2: (parts) => {
                    return <Box className="from-custom-2">{parts}</Box>;
                },
                param: "value from outside"
            }
        );
        renderBox(message);

        const element = await screen.findByTestId("test");
        expect(element).toMatchInlineSnapshot(`
          <div
            data-testid="test"
          >
            Hello, 
            <div
              class="from-custom-1"
            >
              Content of custom tag 
              <div
                class="from-custom-2"
              >
                and nested tag with value from outside
              </div>
            </div>
            !
          </div>
        `);
    });

    it("does not error for static strings", async () => {
        const intl = createPackageIntl("en-US", {
            test: "hi"
        });
        const message = intl.formatRichMessage({
            id: "test"
        });
        const nodes = dumpReactTree(message);
        expect(nodes).toMatchInlineSnapshot(`
          {
            "children": [
              {
                "key": "<no key>",
                "type": "primitive",
                "value": "hi",
              },
            ],
            "key": "<no key>",
            "type": Symbol(react.fragment),
          }
        `);
    });

    it("assigns unique keys to children and nested children", async () => {
        const intl = createPackageIntl("en-US", {
            test: "a <nested2><code>b</code> c <nested1> d <code>e</code> f <code>g</code> h</nested1></nested2>"
        });
        const message = intl.formatRichMessage(
            {
                id: "test"
            },
            {
                nested1(parts) {
                    return <Box className="nested1">{parts}</Box>;
                },
                nested2(parts) {
                    return <Box className="nested2">{parts}</Box>;
                },
                h: "foo"
            }
        );

        renderBox(message);
        const element = await screen.findByTestId("test");
        expect(element).toMatchInlineSnapshot(`
          <div
            data-testid="test"
          >
            a 
            <div
              class="nested2"
            >
              <code>
                b
              </code>
               c 
              <div
                class="nested1"
              >
                 d 
                <code>
                  e
                </code>
                 f 
                <code>
                  g
                </code>
                 h
              </div>
            </div>
          </div>
        `);

        const nodes = dumpReactTree(message);
        expect(nodes).toMatchSnapshot();
    });
});

it("generates a type error when using formatMessage with non-primitive types", async () => {
    const intl = createPackageIntl("en-US", {
        test: "Hello, {name}!"
    });
    const message = intl.formatMessage(
        {
            id: "test"
        },
        {
            // @ts-expect-error Only primitive types are supported
            name: {}
        }
    );

    // This would need some postprocessing to work well in react, which is why formatRichMessage() exists.
    expect(message).toMatchInlineSnapshot(`
    [
      "Hello, ",
      {},
      "!",
    ]
  `);
});

interface TreeNode {
    type: string;
    key: string;
    value?: string;
    children?: TreeNode[];
}

function dumpReactTree(node: any): TreeNode {
    if (node == null || typeof node === "boolean" || typeof node === "string") {
        return {
            type: "primitive",
            key: "<no key>",
            value: node
        };
    }

    const type = node.type || "<no type>";
    const key = node.key || "<no key>";
    const children = node.props?.children;

    const treeNode: TreeNode = {
        type,
        key
    };
    if (children) {
        const childrenArray = Array.isArray(children) ? children : [children];
        treeNode.children = childrenArray.map((child: any) => dumpReactTree(child));
    }
    return treeNode;
}

function renderBox(content: ReactNode) {
    return render(
        <PackageContextProvider>
            <Box data-testid="test">{content}</Box>
        </PackageContextProvider>
    );
}
