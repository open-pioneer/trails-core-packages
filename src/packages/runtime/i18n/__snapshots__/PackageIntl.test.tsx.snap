// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatRichMessage > assigns unique keys to children and nested children 2`] = `
{
  "children": [
    {
      "key": "<no key>",
      "type": "primitive",
      "value": "a ",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "key": "<no key>",
                      "type": "primitive",
                      "value": "b",
                    },
                  ],
                  "key": "<no key>",
                  "type": "code",
                },
              ],
              "key": ".$42/.0",
              "type": Symbol(react.fragment),
            },
            {
              "key": "<no key>",
              "type": "primitive",
              "value": " c ",
            },
            {
              "children": [
                {
                  "children": [
                    {
                      "key": "<no key>",
                      "type": "primitive",
                      "value": " d ",
                    },
                    {
                      "children": [
                        {
                          "children": [
                            {
                              "key": "<no key>",
                              "type": "primitive",
                              "value": "e",
                            },
                          ],
                          "key": "<no key>",
                          "type": "code",
                        },
                      ],
                      "key": ".$42/.1",
                      "type": Symbol(react.fragment),
                    },
                    {
                      "key": "<no key>",
                      "type": "primitive",
                      "value": " f ",
                    },
                    {
                      "children": [
                        {
                          "children": [
                            {
                              "key": "<no key>",
                              "type": "primitive",
                              "value": "g",
                            },
                          ],
                          "key": "<no key>",
                          "type": "code",
                        },
                      ],
                      "key": ".$42/.3",
                      "type": Symbol(react.fragment),
                    },
                    {
                      "key": "<no key>",
                      "type": "primitive",
                      "value": " h",
                    },
                  ],
                  "key": "<no key>",
                  "type": {
                    "$$typeof": Symbol(react.forward_ref),
                    "render": [Function],
                  },
                },
              ],
              "key": ".$42/.2",
              "type": Symbol(react.fragment),
            },
          ],
          "key": "<no key>",
          "type": {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
      ],
      "key": ".$42/.1",
      "type": Symbol(react.fragment),
    },
  ],
  "key": "<no key>",
  "type": Symbol(react.fragment),
}
`;
